contracts/Pool.sol

```diff
@@ -14,260 +15,368 @@ import "./IBorrowersRegistry.sol";
-        uint256 interests = depositedWithInterests - deposits[user];
-        deposits[user] = depositedWithInterests;
-        totalDeposited = totalDeposited + interests;
-        depositIndex.updateUser(user);
-    }
 
+  /**
+   * Returns the current deposited amount for the given user
+   * The value includes the interest rates earned at the current moment
+   * @dev _user the address of queried depositor
+  **/
+  function balanceOf(address user) public view override returns (uint256) {
+    return depositIndex.getIndexedValue(_deposited[user], user);
+  }
 
-    function accumulateBorrowingInterests(address user) internal {
-        uint256 borrowedWithInterests = getBorrowed(user);
-        uint256 interests = borrowedWithInterests - borrowed[user];
-        borrowed[user] = borrowedWithInterests;
-        totalBorrowed = totalBorrowed + interests;
-        borrowIndex.updateUser(user);
-    }
 
+  /**
+   * Returns the current interest rate for deposits
+  **/
+  function getDepositRate() public view returns (uint256) {
+    return _ratesCalculator.calculateDepositRate(totalBorrowed(), totalSupply());
+  }
+
+
+  /**
+   * Returns the current interest rate for borrowings
+  **/
+  function getBorrowingRate() public view returns (uint256) {
+    return _ratesCalculator.calculateBorrowingRate(totalBorrowed(), totalSupply());
+  }
+
+
+  /**
+   * Recovers the residual funds resultant from difference between deposit and borrowing rates
+  **/
+  function recover(uint256 amount, address account) public onlyOwner nonReentrant {
+    uint256 residual = address(this).balance - totalSupply();
+
+    require(amount <= residual, "Trying to recover more residual funds than available");
+
+    payable(account).transfer(amount);
+  }
+
+
+  /* ========== INTERNAL FUNCTIONS ========== */
+
+  function _mint(address account, uint256 amount) internal {
+    require(account != address(0), "ERC20: mint to the zero address");
+
+    _deposited[account] += amount;
+    _deposited[address(this)] += amount;
 
-    /* ========== MODIFIERS ========== */
+    emit Transfer(address(0), account, amount);
+  }
 
-    modifier canBorrow() {
-      require(address(borrowersRegistry) != address(0), "Borrowers registry is not configured.");
-      require(borrowersRegistry.canBorrow(msg.sender), "Only the accounts authorised by borrowers registry may borrow.");
-      _;
-      require(totalDeposited > 0, "Cannot borrow from an empty pool.");
-      require(totalBorrowed * 1 ether / totalDeposited <= MAX_POOL_UTILISATION, "The pool utilisation cannot be greater than 95%.");
+
+  function _burn(address account, uint256 amount) internal {
+    require(account != address(0), "ERC20: burn from the zero address");
+
+    require(_deposited[address(this)] >= amount, "ERC20: burn amount exceeds current pool indexed balance");
+
+    require(_deposited[account] >= amount, "ERC20: burn amount exceeds user balance");
+
+    // verified in "require" above
+    unchecked {
```

Comment:
Is this `unchecked` block necessary?

Reply:
@jakub-wojciechowski

Reply:
It serves optimization since safety check is performed earlier